libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1065,6 +1065,103 @@ string YulUtilFunctions::clearStorageArrayFunction(ArrayType const& _type)
 	});
 }
 
+string YulUtilFunctions::copyArrayToStorage(ArrayType const& _fromType, ArrayType const& _toType)
+{
+	solAssert(
+		*_fromType.copyForLocation(_toType.location(), _toType.isPointer()).get() == dynamic_cast<ReferenceType const&>(_toType),
+		""
+	);
+
+	string functionName = "copy_array_to_storage_from_" + _fromType.identifier() + "_to_" + _toType.identifier();
+	return m_functionCollector.createFunction(functionName, [&](){
+		Whiskers templ(R"(
+			function <functionName>(slot, <value>) {
+				let length := <arrayLength>(<value>)
+				<?isToDynamic>
+					<storeLength>(slot, length)
+				</isToDynamic>
+
+				for { let i := 0 } lt(i, length) {i := add(i, 1)} {
+					let callDataOrMemoryOffset := <fromArrayIndexAccess>(<value>, i)
+					let elementSlot, elementOffset := <toArrayIndexAccess>(slot, i)
```

Comment:
I'm wondering... both of these functions are **very** expensive, because they do all kinds of bounds checks and verifications...
All of them could be done only once in advance and then inside the loop it would basically just be incrementing an offset by a stride...

I'd like to think the optimizer to be clever enough to inline these, recognize that the checks hold inductively, if and only if they hold for "i = length - 1", extract this check in front of the loop and remove all checks inside the loop... but that's never going to happen :-)...

Reply:
I removed `fromArrayIndexAccess`, am at half of removing `toArrayIndexAccess`, will update with that hopefully today